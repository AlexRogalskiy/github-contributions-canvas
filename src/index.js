import moment from "moment";

function getDateInfo(data, date) {
  return data.contributions.find(contrib => contrib.date === date);
}

function getContributionCount(graphEntries) {
  return graphEntries.reduce((rowTotal, row) => {
    return (
      rowTotal +
      row.reduce((colTotal, col) => {
        return colTotal + (col.info ? col.info.count : 0);
      }, 0)
    );
  }, 0);
}

const DATE_FORMAT = "YYYY-MM-DD";
const boxWidth = 10;
const boxMargin = 2;
const textHeight = 15;
const fontFace = "IBM Plex Mono";
const headerHeight = 60;
const canvasMargin = 20;
const yearHeight = textHeight + (boxWidth + boxMargin) * 7 + canvasMargin;
const scaleFactor = window.devicePixelRatio || 1;

function drawYear(ctx, year, offsetX = 0, offsetY = 0, data) {
  const thisYear = moment().format("YYYY");
  const today = year.year === thisYear ? moment() : moment(year.range.end);
  const start = moment(`${year.year}-01-01`);
  const firstDate = start.clone();

  if (firstDate.day() !== 6) {
    firstDate.day(-(firstDate.day() + 1 % 7));
  }

  const nextDate = firstDate.clone();
  const firstRowDates = [];
  const graphEntries = [];

  while (nextDate <= today && nextDate.day(7) <= today) {
    const date = nextDate.format(DATE_FORMAT);
    firstRowDates.push({
      date,
      info: getDateInfo(data, date)
    });
  }

  graphEntries.push(firstRowDates);

  for (let i = 1; i < 7; i += 1) {
    graphEntries.push(
      firstRowDates.map(dateObj => {
        const date = moment(dateObj.date)
          .day(i)
          .format(DATE_FORMAT);
        return {
          date,
          info: getDateInfo(data, date)
        };
      })
    );
  }

  const count = new Intl.NumberFormat().format(
    getContributionCount(graphEntries)
  );

  ctx.textBaseline = "hanging";
  ctx.fillStyle = "#000000";
  ctx.font = `10px '${fontFace}'`;
  ctx.fillText(
    `${year.year}: ${count} Contribution${year.total === 1 ? "" : "s"}${
      thisYear === year.year ? " (so far)" : ""
    }`,
    offsetX,
    offsetY
  );

  for (let y = 0; y < graphEntries.length; y += 1) {
    for (let x = 0; x < graphEntries[y].length; x += 1) {
      const day = graphEntries[y][x];
      if (moment(day.date) > today || !day.info) {
        continue;
      }
      ctx.fillStyle = day.info.color;
      ctx.fillRect(
        offsetX + (boxWidth + boxMargin) * x,
        offsetY + textHeight + (boxWidth + boxMargin) * y,
        10,
        10
      );
    }
  }
}

function drawMetaData(ctx, opts = {}) {
  const { username, width, height, footerText } = opts;
  ctx.fillStyle = "#ffffff";
  ctx.fillRect(0, 0, width, height);

  if (footerText) {
    ctx.fillStyle = "#666666";
    ctx.textBaseline = "bottom";
    ctx.font = `10px '${fontFace}'`;
    ctx.fillText(footerText, canvasMargin, height - 5);
  }

  ctx.fillStyle = "#000000";
  ctx.textBaseline = "hanging";
  ctx.font = `20px '${fontFace}'`;
  ctx.fillText(`@${username} on Github`, canvasMargin, canvasMargin);

  ctx.beginPath();
  ctx.moveTo(canvasMargin, 55);
  ctx.lineTo(width - canvasMargin, 55);
  ctx.strokeStyle = "#EBEDF0";
  ctx.stroke();
}

export function drawContributions(canvas, opts) {
  const { data, username } = opts;
  const height =
    data.years.length * yearHeight + canvasMargin + headerHeight + 10;
  const width = 53 * (boxWidth + boxMargin) + canvasMargin * 2;

  canvas.width = width * scaleFactor;
  canvas.height = height * scaleFactor;

  const ctx = canvas.getContext("2d");
  ctx.scale(scaleFactor, scaleFactor);
  ctx.textBaseline = "hanging";

  drawMetaData(ctx, {
    ...opts,
    width,
    height
  });

  data.years.forEach((year, i) => {
    const offsetY = yearHeight * i + canvasMargin + headerHeight;
    const offsetX = canvasMargin;
    drawYear(ctx, year, offsetX, offsetY, data);
  });
}
