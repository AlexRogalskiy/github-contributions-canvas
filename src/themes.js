export const themes = {
  __test__: {
    background: "#ffffff",
    text: "#ffffff",
    meta: "#ffffff",
    grade4: "#196127",
    grade3: "#239a3b",
    grade2: "#7bc96f",
    grade1: "#c6e48b",
    grade0: "#ebedf0"
  },
  standard: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#196127",
    grade3: "#239a3b",
    grade2: "#7bc96f",
    grade1: "#c6e48b",
    grade0: "#ebedf0"
  },
  halloween: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#03001C",
    grade3: "#FE9600",
    grade2: "#FFC501",
    grade1: "#FFEE4A",
    grade0: "#ebedf0"
  },
  teal: {
    background: "#ffffff",
    text: "#000000",
    meta: "#666666",
    grade4: "#458B74",
    grade3: "#66CDAA",
    grade2: "#76EEC6",
    grade1: "#7FFFD4",
    grade0: "#ebedf0"
  },
  leftPad: {
    background: "#000000",
    text: "#ffffff",
    meta: "#999999",
    grade4: "#F6F6F6",
    grade3: "#DDDDDD",
    grade2: "#A5A5A5",
    grade1: "#646464",
    grade0: "#2F2F2F"
  },
  dracula: {
    background: "#282a36",
    text: "#f8f8f2",
    meta: "#44475a",
    grade4: "#ff5555",
    grade3: "#ff79c6",
    grade2: "#bd93f9",
    grade1: "#ffb86c",
    grade0: "#f1fa8c"
  },
  blue: {
    background: "#181818",
    text: "#C0C0C0",
    meta: "#666666",
    grade4: "#4F83BF",
    grade3: "#416895",
    grade2: "#344E6C",
    grade1: "#263342",
    grade0: "#222222"
  },
  panda: {
    background: "#2B2C2F",
    text: "#E6E6E6",
    meta: "#676B79",
    grade4: "#FF4B82",
    grade3: "#19f9d8",
    grade2: "#6FC1FF",
    grade1: "#34353B",
    grade0: "#242526"
  }
};
