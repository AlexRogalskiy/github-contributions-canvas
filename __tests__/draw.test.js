import mockData from "./mock.json";
import { drawContributions } from "../src";
import { createCanvas } from "canvas";
const canvas = createCanvas(200, 200);

test("draws contributions correctly", () => {
  drawContributions(canvas, {
    data: mockData,
    username: "sallar",
    footerText: "test",
    theme: "standard"
  });
  expect(canvas.toDataURL()).toMatchSnapshot();
});
