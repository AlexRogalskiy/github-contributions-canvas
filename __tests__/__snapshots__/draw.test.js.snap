// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draws contributions correctly 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqQAAAQSCAYAAAB5Bej7AAAABmJLR0QA/wD/AP+gvaeTAAAciklEQVR4nO3dsWoc32OGYTm4SNIYjCu5daErMBh8NQL3KlyqVOlCvUFXYzDoClS4tSpjcJOk2zQpwq8wM/rP2feM9nnqj2W02pVeTnNeHA6HwxkAAET+rX4AAABOmyAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgNTLrV/wv/77f7Z+SQAAJvKf//Hvm76eE1IAAFIvDofDoX4IAABOlxNSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgNTLrV/w+88vi3Yf3n4+yv7X7z+L9m9evzrK81zfXy3a37y/PcrzzPZ+rn1/1j7P6Ndf+/O++/px0f7Hp2+LdgCwR05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi9OBwOh/ohAAA4XU5IAQBIPfu77EffLT76rva1+/PLi0X7x7uHJ73+bL+vU3t+AHiOnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACkNr/LfvRd8KPvCl/7/KPvph/9PGtf/9fvP4v2b16/etJ+ts/D6OdZ+/4DwHPkhBQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlLvsAQBIOSEFACCV32U/293io59n7z+vu+z//vprf97zy4tF+8e7h0U7ANgjJ6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAACpze+yH30X+ej9u68fF+1/fPr2pNdfe5f62v3o55/trvnRP+/au+nX7tc+PwA8R05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUpvfZT/6rva1+7V3i699/dF3tc/2/Ke2H/3+AwBOSAEAiAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIucseAICUE1IAAFKb32U/293io59n7d3lo++a3/td9qPvjp/teQAAJ6QAAMQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApF5u/YLff35ZtPvw9vNJ7q/vrxbtb97fnp2dnZ39+v1n0f7N61dPev3Zft7Znufd14+L9j8+fTvK85za53nv7+fef961zz96P9v7P/p5Zvs8r92v/fs5+u+Jz8Pfn2ftfmtOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIvTgcDof6IQAAOF1OSAEASG1+l/3ou1Znu8t1tv3a92f0fvRdxrPdTT/b73fvrz/b73e292f03eKj79re+/dx9N3ls32eZ3ue2X6/s32e137eak5IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUru7y362/ei7nmf7eWd7/rV3B892F/no559tP/r9PLX9bHfZ7/3vyWz7vd8dP/rv4fnlxaL9493Dk55n7c+79nlm+33VnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACkNr/Lfra7WU9tP9td2LPdXT7b+z/6/Zzt8zPb72u23+/oz//ou+xn+/2uff7R+9nen9m+77M9z+j/j2s/P6M/DzUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAAKn8Lnv7bfdr79I9v7xYtH+8e3jSfra7uU/tbvG93z09+vc1+vXXfl9m+3sy213toz//s90tPttd9rO9P6O/j6f2ea45IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEi5y/7I+7V3y852l/Tau79ne/9nu1t5truPR/9+R78/o3/etfvZvo9r97P9/ZntLvXZnmfvd9nv/fs4+v0c/fe55oQUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJS77AEASDkhBQAg5S77f+z3ftf8qe3Xvp+z3WU8+m7iU3v/976f7W700c9zaq8/+v/F6P1sn8/Rzz/bzzt6X3NCCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDKXfYAAKSckAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDq5dYv+Ov3n0W7N69fnZ2dnZ1d318t2t+8vz3K659fXizaP949HOX11/68339+WbT/8PbzUfajn3/t+7l2P/rntd92P9v38d3Xj4v2Pz59Ozs7m+/7OPrv82zPM9v7P9v7s/b1Z/v/u/b7OPr5Z/t/vfb92ZoTUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSLw6Hw6F+CAAATpcTUgAAUpvfZb/3u4BH3z3tLua/70e/n7PdHTzbXfB738/2eZjtefb++5rt+Wd7/0c/z97vst/793ftfu37ufb1t+aEFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUu+wBAEg5IQUAILX5Xfaz3TU8213qsz3/qd3FPNvPa7/tfu+/39Hf33dfPy7a//j07UmvP/pu9FP7/c521/zoz9vo/Wzfx9G9Mfr3uzUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAAKn8LvvRd62O3p9fXizaP949POn1937X82x308/2/thvu5/t+zX6eU7t+7L3vyd7f/61/+9G370+2/7d14+L9j8+fXvS64/+e7J2vzUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAAKn8LvvRd/XOdrfsbO/PbHclz/Y8s/1+T20/+m732b7vs91tPdvnYe/fx9n+Hs52V/vo5xn993/088z2827NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEBqd3fZj96fX14s2j/ePZydnc13t/JszzPbXedr7wKe7ecd/Xk4tc/PbJ/P0c8z213Yp/b3bbb3c7bnX/v5HP33f7bvy+j92vdza05IAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEi5yx4AgJQTUgAAUru7y370XfOj737d+13Yp3bX9mx3Se99P/r7Ndtd6ms/z7PdLT76789szz/b9330+zP6/+na99/nbV/PszUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkXm79gr9+/1m0e/P6lb29/Yntv//8smj/4e3nozyP/fPev/v6cdH+x6dvT3r96/urRfub97dPev3Zvi+zPc/o3+/a1z+/vFi0f7x7eNLzjP68rd1vzQkpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAKkXh8PhUD8EAACnywkpAAApd9nb22+4n+2u59H7tXcrz/b+rH2e2Z7fftv9qf1+1/68o+9Sn+39P7W/D2uff2tOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIucseAICUE1IAAFLusre3t592f2p3Z9tvu3/39eOi/Y9P3560n+3nne37tff93u+yH/08W3NCCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEDKXfYAAKSckAIAkHKXvb39X/buev77fu1dyad2V7i9vf08+9F30892l/3o/19bc0IKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQMpd9gAApJyQAgCQcpf9P/bu2rY/5f355cWi/ePdw5Nef7a7nmfbe3+23Y++y/vU9j6f+9qP/nu+NSekAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKTcZQ8AQMoJKQAAKXfZ/2Pv7mP7U96Pvqt69Ov7/tr//7271/e19/fn768/+nlqTkgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASLnLHgCAlBNSAABSu7vLfu933e597/153vu1v9+1dyX7/v59P9td2Hvfn19eLNo/3j0c5Xn2vh/9+Zzt78Nsf99G72tOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIucseAICUE1IAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIvdz6BX/9/rNo9+b1qyftzy8vFu0f7x6O8jz29vb2W+39fbO3t9/LfmtOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBIvTgcDof6IQAAOF1OSAEASO3uLnt7+1PeX99fLdrfvL89yvOc2t77b29vb+8uewAAniFBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb29vb29vb29var9ltzQgoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBAyl32AACknJACAJByl729/Yb76/urRfub97dHeR57e3t7e/sR+605IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEi5y97e3t7e3t7e/sT3NSekAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKTcZQ8AQMoJKQAAKXfZ20+1v76/WrS/eX97lOext7e3t7c/hX3NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb29vb29vb29pPvr++vFu1v3t8+6fVrTkgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASLnLHgCAlBNSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASL3c+gV//f6zaPfm9Sv7E9h///ll0f7D289HeR57e/tx+9m+77M9j/3f99f3V4v2N+9vj/I89n/fb80JKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAACpF4fD4VA/BAAAp8sJKQAAqWd/l/3au4zdfWxvb2//PPb+nu9r7/e1r/3WnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACk3GX/f9xlb29vb29vb2+/bL81J6QAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApNxlDwBAygkpAACpZ3+Xvb29vb29vb29/d/3NSekAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKTcZQ8AQMoJKQAAKXfZ20+1//7zy6L9h7efj/I89vb24/bnlxeL9o93D0d5Hnv7U97XnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACknv1d9u++fly0//Hp21Gex97e3n6r/fX91aL9zfvbozzP6P33n18W7T+8/XyU5zm/vFi0f7x7OMrz2Nv/K/uaE1IAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUu6yBwAg5YQUAIDUs7/L3t7e3t7efsT++88vi/Yf3n4+yvPY2/8r+5oTUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABS7rIHACDlhBQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSL7d+wV+//yzavXn9yt7e3t7e3t7efof7rTkhBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACD14nA4HOqHAADgdDkhBQAg5S57e3t7e3t7e3v7VfutOSEFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAIOUuewAAUk5IAQBIucve3t7e3t7e3v7E9zUnpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACk3GUPAEDKCSkAACl32dvb29vb2z+D/fnlxaL9493DUZ7Hfl/7mhNSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFLusgcAIOWEFACAlLvs7afau4vZ3t7e3n7J/vr+atH+5v3tUZ5n7/uaE1IAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUu6yBwAg5YQUAIDU5nfZu4vc3t7e3v4U9rPdpe7/r/2/sq85IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg5S57AABSTkgBAEhtfpf9bHez2tvb29vb29vb/31fc0IKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQMpd9gAApJyQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkBKkAACkBCkAAClBCgBASpACAJASpAAApAQpAAApQQoAQOrl1i/46/efRbs3r1/Z29vb29vb29tPsK85IQUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAg9eJwOBzqhwAA4HQ5IQUAIOUue3t7e3t7e3v7E9/XnJACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKUEKAEBKkAIAkHKXPQAAKSekAACk3GVvb29vb29vb3/i+5oTUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABS7rIHACDlhBQAgJS77O3tw/355cWi/ePdw1Gex/5577///LJo/+Ht56M8j729/Tz7mhNSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFLusgcAIOWEFACAlLvs7e3D/buvHxftf3z6dpTnsbe3t7c/zX3NCSkAAClBCgBASpACAJASpAAApAQpAAApQQoAQEqQAgCQEqQAAKQEKQAAKXfZAwCQckIKAEDKXfb29uH++88vi/Yf3n4+yvPY2z+n/fnlxaL9493DUZ7Hfl/7tX+fr++vFu1v3t8+6XncZQ8AAAMJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASLnLHgCAlBNSAABS7rK3tw/37rK3tx+39/2yt1++rzkhBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACDlLnsAAFJOSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAICUIAUAICVIAQBICVIAAFKCFACAlCAFACAlSAEASAlSAABSghQAgJQgBQAgJUgBAEgJUgAAUoIUAIDU/wJDF4ZV3BaPOwAAAABJRU5ErkJggg=="`;
